Feat/rpc limiter common

IvanBelyakoff opened the pull request:
Implemented limiter for transfers `history` RPC requests (it does not limit for now any other RPC calls. Though it is easily possible to do if needed)

Limits for transfer history:
- 5k requests per account (totally for all history, not per day)
- 10k requests totally per day

Since we should allow requests to blockchain to detect new transactions, we should differentiate those with history requests. For that purpose I added tags to rpc.Client:
- tag (used to mark history calls per account)
- groupTag (used to limit total history calls)

Other improvements and things to mention:
- ClientInterface split to ChainInterface and others for better clarity
- Fixed RPC stats to properly display tagged RPC calls
- RPCLimiter renamed to RPCRpsLimiter, as there are 2 limiters now
- Replaced some fields and corresponding locks to atomics and `sync` types where needed since chain.Client is now copied
- Used connection.Connectable for ClientInterface

Closes #5173

IvanBelyakoff committed:
feat(wallet)_: Implemented RPC limiter interface

Split ClientInterface to aggregation of multiple interfaces
Added tags to RPC stats API
Use tagged RPC client for transfers commands
Implemented general interface for RPC limiting

IvanBelyakoff committed:
feat(wallet)_: add limiter to rpc.client

Add tests for limiter
Use limiter for transfer history commands

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (16)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_multiplication_x: | 1dc1f06d | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/1/) | 2024-05-21 04:56:02 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250 | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5194/1/) | 2024-05-21 04:58:04 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240521-045310-1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250-pr5194.zip) |
| :heavy_check_mark: | 1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250 | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5194/1/) | 2024-05-21 04:58:24 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240521-045310-1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250-pr5194.zip) |
| :heavy_check_mark: | 1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250 | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5194/1/) | 2024-05-21 04:59:04 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240521-045306-1dc1f06da42d68b5a6ff94cc3cd92b7cfe2a5250-pr5194.aar) |
| | | | | | | |
| :heavy_multiplication_x: | 2fe64f62 | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/2/) | 2024-05-22 07:40:12 | ~2 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/2//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 2fe64f62df62c1a6f373b508d76fb00783992334 | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5194/2/) | 2024-05-22 07:40:19 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240522-073728-2fe64f62df62c1a6f373b508d76fb00783992334-pr5194.aar) |
| :heavy_check_mark: | 2fe64f62df62c1a6f373b508d76fb00783992334 | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5194/2/) | 2024-05-22 07:40:57 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240522-073728-2fe64f62df62c1a6f373b508d76fb00783992334-pr5194.zip) |
| :heavy_check_mark: | 2fe64f62df62c1a6f373b508d76fb00783992334 | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5194/2/) | 2024-05-22 07:41:39 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240522-073732-2fe64f62df62c1a6f373b508d76fb00783992334-pr5194.zip) |
| | | | | | | |
| :heavy_check_mark: | 7c4fb95885714f78a007485be473aaa2efc34ce2 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5194/3/) | 2024-05-22 10:21:15 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240522-101829-7c4fb95885714f78a007485be473aaa2efc34ce2-pr5194.aar) |
| :heavy_check_mark: | 7c4fb95885714f78a007485be473aaa2efc34ce2 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5194/3/) | 2024-05-22 10:21:24 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240522-101829-7c4fb95885714f78a007485be473aaa2efc34ce2-pr5194.zip) |
| :heavy_check_mark: | 7c4fb95885714f78a007485be473aaa2efc34ce2 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5194/3/) | 2024-05-22 10:22:16 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240522-101829-7c4fb95885714f78a007485be473aaa2efc34ce2-pr5194.zip) |
| | | | | | | |
| :heavy_check_mark: | 32376bf7b9d7988b0b77a9268457e2aaa9e1e88d | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5194/4/) | 2024-05-22 10:26:54 | ~3 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240522-102348-32376bf7b9d7988b0b77a9268457e2aaa9e1e88d-pr5194.zip) |
| :heavy_check_mark: | 32376bf7b9d7988b0b77a9268457e2aaa9e1e88d | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5194/4/) | 2024-05-22 10:27:55 | ~4 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240522-102348-32376bf7b9d7988b0b77a9268457e2aaa9e1e88d-pr5194.zip) |
| :heavy_check_mark: | 32376bf7b9d7988b0b77a9268457e2aaa9e1e88d | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5194/4/) | 2024-05-22 10:29:38 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240522-102348-32376bf7b9d7988b0b77a9268457e2aaa9e1e88d-pr5194.aar) |
| | | | | | | |
| :heavy_check_mark: | 7c4fb958 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/3/) | 2024-05-22 11:02:54 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/3//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 32376bf7 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/4/) | 2024-05-22 11:46:17 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/4//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | d341cd216c79a30378552adc119723d271b80f06 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5194/5/) | 2024-05-23 10:05:44 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240523-100302-d341cd216c79a30378552adc119723d271b80f06-pr5194.zip) |
| :heavy_check_mark: | d341cd216c79a30378552adc119723d271b80f06 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5194/5/) | 2024-05-23 10:05:50 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240523-100302-d341cd216c79a30378552adc119723d271b80f06-pr5194.aar) |
| :heavy_check_mark: | d341cd216c79a30378552adc119723d271b80f06 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5194/5/) | 2024-05-23 10:06:42 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240523-100302-d341cd216c79a30378552adc119723d271b80f06-pr5194.zip) |
| | | | | | | |
| :heavy_check_mark: | d341cd21 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/5/) | 2024-05-23 10:46:33 | ~43 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5194/5//consoleText) |

IvanBelyakoff committed:
feat(wallet)_: added and fixed tests for findBlocksCommand with limiter.

IvanBelyakoff committed:
feat(wallet)_: renamed some methods for clarity

IvanBelyakoff committed:
feat(wallet)_: added group tag for RPC chain client

It is needed to be able to set common limits for chain client
Added a test for group tag limiter
Added a mutex to RPC limiter to change counters atomically
Replaced isConnected with atomic.Bool and made it a pointer to
be shared across client instances

IvanBelyakoff committed:
feat(wallet)_: implemented storing RPC requests in sql table

Fixed some tests and minor issues

IvanBelyakoff committed:
feat(wallet)_: fixed lint issues. Fixed rpc stats to include tagged
requests

IvanBelyakoff committed:
feat(wallet)_: Fixed account limit period for history to infinite instead of per day

Fix rpc stats doubling.
Add test for infinite limit

IvanBelyakoff committed:
feat(wallet)_: more fixes for rpc_limiter_db and chain client, more
tests

saledjenic reviewed (approved):
@IvanBelyakoff changes look good, though I haven't tested this @J-Son89 could you or anybody else from the mobile time review this as well?

So in this PR we're limiting a user to:
- 5k requests per account (totally for all history, not per day)
- 10k requests totally per day

After the user spends those calls won't be able to interact with the app (send tx, add new acc and so on...)

J-Son89 commented:
> @IvanBelyakoff changes look good, though I haven't tested this @J-Son89 could you or anybody else from the mobile time review this as well?
> 
> 
> 
> So in this PR we're limiting a user to:
> 
> - 5k requests per account (totally for all history, not per day)
> 
> - 10k requests totally per day
> 
> 
> 
> After the user spends those calls won't be able to interact with the app (send tx, add new acc and so on...)

Sure @saledjenic we'll get some eyes on it! 🙏

J-Son89 commented:
> After the user spends those calls won't be able to interact with the app (send tx, add new acc and so on...)

so a question, if the user hits a limiti for their history etc they won't be able to make transfers that day.
Is the user hitting the limit of these requests within their control?
i.e to hit the limit they have to actively search through the graph etc? 
Because if this limit will just be hit by a user with a large account and logging in then we are kind of punishing that type of user.
Perhaps there needs to be some UX in there to communicate this to the user that they are approaching the limit etc so they understand what is happening. 
Not for this pr of course, but seems like something we should communicate with design? (or perhaps this has already happened? 🤔 )
cc @benjthayer

IvanBelyakoff commented:
> > After the user spends those calls won't be able to interact with the app (send tx, add new acc and so on...)
> 
> so a question, if the user hits a limiti for their history etc they won't be able to make transfers that day. Is the user hitting the limit of these requests within their control? i.e to hit the limit they have to actively search through the graph etc? Because if this limit will just be hit by a user with a large account and logging in then we are kind of punishing that type of user. Perhaps there needs to be some UX in there to communicate this to the user that they are approaching the limit etc so they understand what is happening. Not for this pr of course, but seems like something we should communicate with design? (or perhaps this has already happened? 🤔 ) cc @benjthayer

Jamie, @saledjenic  this limit is only per transfers history loading. So it doe snot affect sending Tx, detection of new transfers, all other not transfers history related things. So user should be able to use the app normally, just if they have big account history or added some watch only accounts with big history, not all history will be loaded for those. But this was the intention of the issue - to limit that history calls.

> > Is the user hitting the limit of these requests within their control?
> i.e to hit the limit they have to actively search through the graph etc?

It is out of their control and those requests will start once they add an account - we start searching for transfers and load them, but as I mentioned above, this is only transfers history that is limited

IvanBelyakoff commented:
> After the user spends those calls won't be able to interact with the app (send tx, add new acc and so on...)

Sale user will be able to interact with the app normally. I'd appreciate some help with testing it from QA team @anastasiyaig because despite the tests I wrote to ensure the limiter should work as expected, in reality I was not able to reach those limits due to some other issues (at least, I saw some rate limiting errors)
